src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -867,6 +867,95 @@ static void Main()
         }
 
         [Fact]
+        public void HugeTupleCreationParses()
+        {
+            StringBuilder b = new StringBuilder();
+            b.Append("(");
+            for (int i = 0; i < 3000; i++)
+            {
+                b.Append("1, ");
+            }
+            b.Append("1)");
+
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        var x = " + b.ToString() + @";
+    }
+}
+";
+            CreateCompilationWithMscorlib(source);
+        }
+
+        [Fact]
+        public void HugeTupleDeclarationParses()
+        {
+            StringBuilder b = new StringBuilder();
+            b.Append("(");
+            for (int i = 0; i < 3000; i++)
+            {
+                b.Append("int, ");
+            }
+            b.Append("int)");
+
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        " + b.ToString() + @" x;
+    }
+}
+";
+            CreateCompilationWithMscorlib(source);
+        }
+
+        [Fact]
+        public void GenericTupleWithoutTupleLibrary()
+        {
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        var x = M<int, bool>();
+        System.Console.WriteLine($""{x.first} {x.second}"");
+    }
+
+    (T1 first, T2 second) M<T1, T2>()
+    {
+        return (default(T1), default(T2));
+    }
+}
+";
+            // PROTOTYPE this should not throw
```

Comment:
I'll change to PROTOTYPE(tuples).

The test itself doesn't throw, but `CompileAndVerify` does (the exception is caught by `Assert.Throws`. Is that ok?